dask/dataframe/tests/test_io.py

```diff
@@ -820,15 +832,88 @@ def test_to_csv():
+            result.index.name = None
             tm.assert_frame_equal(result, df)
 
-    a = dd.from_pandas(df, npartitions)
-    with tmpfile('csv') as fn:
-        r = a.to_csv(fn, compute=False)
-        r.compute()
-        result = pd.read_csv(fn, index_col=0)
-        tm.assert_frame_equal(result, df)
+        with tmpfile('csv') as fn:
+            r = a.to_csv(fn, compute=False)
+            r.compute()
+            result = dd.read_csv(fn).set_index('Unnamed: 0').compute()
+            result.index.name = None
+            eq(result, df)
+
+        with tmpdir() as dn:
+            fn = os.path.join(dn, 'data_*.csv')
+            a.to_csv(fn)
+            result = dd.read_csv(fn).set_index('Unnamed: 0').compute()
+            result.index.name = None
+            eq(result, df)
+
+    a = dd.from_pandas(df, 2)
+    with tmpdir() as dn:
+        with pytest.raises(ValueError):
+            fn = os.path.join(dn, "data_*_*.csv")
+            a.to_csv(fn)
+
+    df16 = pd.DataFrame({'x': ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p'],
+                       'y': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]},
+                            index=[1., 2., 3., 4., 5., 6., 7., 8., 9., 10., 11., 12., 13., 14., 15., 16.])
+    a = dd.from_pandas(df16, 16)
+    with tmpdir() as dn:
+        fn = os.path.join(dn, 'data_*.csv')
+        a.to_csv(fn)
+        result = dd.read_csv(fn).set_index('Unnamed: 0').compute()
+        result.index.name = None
+        eq(result, df16)
+
+    # test handling existing files when links are optimized out
+    a = dd.from_pandas(df, 2)
+    with tmpdir() as dn:
+        fn = os.path.join(dn, 'data_1.csv')
+        a.to_csv(fn)
+        fn = os.path.join(dn, 'data_*.csv')
+        a.to_csv(fn, mode='w')
+        result = dd.read_csv(fn).set_index('Unnamed: 0').compute()
+        result.index.name = None
+        eq(result, df)
+
+    # test handling existing files when links are optimized out
+    a = dd.from_pandas(df16, 16)
+    with tmpdir() as dn:
+        fn = os.path.join(dn, 'data_01.csv')
+        a.to_csv(fn)
+        fn = os.path.join(dn, 'data_*.csv')
+        a.to_csv(fn, mode='w')
+        result = dd.read_csv(fn).set_index('Unnamed: 0').compute()
+        result.index.name = None
+        eq(result, df16)
+
+    # test handling existing files when links are optimized out
+    df = pd.DataFrame({'x': ['a', 'b', 'c', 'd'],
+                       'y': [1, 2, 3, 4]}, index=[1., 2., 3., 4.])
+    a = dd.from_pandas(df, 2)
+    with tmpdir() as dn:
+        with pytest.raises(ValueError):
+            fn = os.path.join(dn, 'data_1.csv')
+            a.to_csv(fn)
+            fn = os.path.join(dn, 'data_*.csv')
+            a.to_csv(fn, mode='a')
+            result = dd.read_csv(fn).set_index('Unnamed: 0').compute()
+            result.index.name = None
+            eq(result, df.append(df))
```

Comment:
Since this error is manually raised, I wanted it there for completeness sake. In case anybody tries supporting that case, he'll have a basic test ready.